File: geotracking/recording/cpp/src/gnss_reader/GPSDReader.cpp
#include "gnss_reader/GPSDReader.hpp"

GPSDReader::GPSDReader(){
    
}

GPSDReader::~GPSDReader()
{
    continue_to_grab = false;
    grab_gnss_data.join();
#ifdef GPSD_FOUND
    
#else
    std::cerr << "[library not found] GPSD library was not found ... please install it before using this sample" << std::endl;
#endif
}
void GPSDReader::initialize()
{
    std::cout << "initialize " << std::endl;
    grab_gnss_data = std::thread(&GPSDReader::grabGNSSData, this);
#ifdef GPSD_FOUND
    std::cout << "Create new object" << std::endl;
    gnss_getter.reset(new gpsmm("localhost", DEFAULT_GPSD_PORT));
    if (gnss_getter->stream(WATCH_ENABLE | WATCH_JSON) == nullptr)
    {
        std::cerr << "No GPSD running .. exit" << std::endl;
        exit(EXIT_FAILURE);
    }
    std::cout << "Successfully opened GPS stream" << std::endl;
    std::cout << "Waiting for GNSS fix" << std::endl;

    bool received_fix = false;
    struct gps_data_t *gpsd_data;
    while (!received_fix)
    {
        if (!gnss_getter->waiting(0))
            continue;
        if ((gpsd_data = gnss_getter->read()) == NULL)
        {
            std::cerr << "[GNSS] read error ... exit program" << std::endl;
            exit(EXIT_FAILURE);
        }
        if (gpsd_data->fix.mode >= MODE_2D)
            received_fix = true;
    }
    is_initialized_mtx.lock();
    is_initialized = true;
    is_initialized_mtx.unlock();
    std::cout << "Fix found !!!" << std::endl;
#else
    std::cerr << "[library not found] GPSD library was not found ... please install it before using this sample" << std::endl;
#endif
}

sl::GNSSData GPSDReader::getNextGNSSValue()
{
#ifdef GPSD_FOUND
    // 0. Check if GNSS is initialized:
    // 1. Get GNSS datas:
    struct gps_data_t *gpsd_data;
    while ((gpsd_data = gnss_getter->read()) == NULL)
        ;
    if (gpsd_data->fix.mode >= MODE_2D)
    {
        sl::GNSSData current_gnss_data;
        // Fill out coordinates:
        current_gnss_data.setCoordinates(gpsd_data->fix.latitude, gpsd_data->fix.longitude, gpsd_data->fix.altMSL, false);
        // Fill out default standard deviation:
        current_gnss_data.longitude_std = current_gnss_data.latitude_std = 0.001f;
        current_gnss_data.altitude_std = 1.f;
        // Fill out covariance [must be not null]
        std::array<double, 9> position_covariance;
        position_covariance[0] = gpsd_data->fix.eph * gpsd_data->fix.eph;
        position_covariance[1 * 3 + 1] = gpsd_data->fix.eph * gpsd_data->fix.eph;
        position_covariance[2 * 3 + 2] = gpsd_data->fix.epv * gpsd_data->fix.epv;
        current_gnss_data.position_covariance = position_covariance;
        // Compute timestamp
        uint64_t current_ts_gps = gpsd_data->fix.time.tv_sec * 1000000;
        uint64_t current_tns_gps = gpsd_data->fix.time.tv_nsec / float(1000);
        auto current_gnss_timestamp = current_ts_gps + current_tns_gps;
        current_gnss_data.ts.setMicroseconds(current_gnss_timestamp);


        return current_gnss_data;
    }
    else
    {
        std::cout << "Fix lost: reinit GNSS" << std::endl;
        initialize();
        return getNextGNSSValue();
    }
#else
    std::cerr << "[library not found] GPSD library was not found ... please install it before using this sample" << std::endl;
#endif

    return sl::GNSSData();
}

sl::ERROR_CODE GPSDReader::grab(sl::GNSSData & current_data){
    if(new_data){
        new_data=false;
        current_data = current_gnss_data;
        return sl::ERROR_CODE::SUCCESS;
    }
    return sl::ERROR_CODE::FAILURE;
}

void GPSDReader::grabGNSSData(){
    while(1){
        is_initialized_mtx.lock();
        if(is_initialized){
            is_initialized_mtx.unlock();
            break;
        }
        is_initialized_mtx.unlock();
        std::this_thread::sleep_for(std::chrono::milliseconds(1));
    }
    while (continue_to_grab)
    {
        #ifdef GPSD_FOUND
        current_gnss_data = getNextGNSSValue();
        new_data = true;
        #endif
    }
    
}

File: global localization/recording/python/gnss_reader/gpsd_reader.py
import threading
import time
import pyzed.sl as sl
from gpsdclient import GPSDClient
import random
import datetime


class GPSDReader:
    def __init__(self):
        self.continue_to_grab = True
        self.new_data = False
        self.is_initialized = False
        self.current_gnss_data = None
        self.is_initialized_mtx = threading.Lock()
        self.client = None
        self.gnss_getter = None
        self.grab_gnss_data = None

    def initialize(self):
        try:
            self.client = GPSDClient(host="127.0.0.1")
        except:
            print("No GPSD running .. exit")
            return -1

        self.grab_gnss_data = threading.Thread(target=self.grabGNSSData)
        self.grab_gnss_data.start()
        print("Successfully connected to GPSD")
        print("Waiting for GNSS fix")
        received_fix = False

        self.gnss_getter = self.client.dict_stream(convert_datetime=True, filter=["TPV"])
        while not received_fix:
            gpsd_data = next(self.gnss_getter)
            if "class" in gpsd_data and gpsd_data["class"] == "TPV" and "mode" in gpsd_data and gpsd_data["mode"] >= 2:
                received_fix = True
        print("Fix found !!!")
        with self.is_initialized_mtx:
            self.is_initialized = True
        return 0

    def getNextGNSSValue(self):
        gpsd_data = None
        while gpsd_data is None:
            gpsd_data = next(self.gnss_getter)

        if "class" in gpsd_data and gpsd_data["class"] == "TPV" and "mode" in gpsd_data and gpsd_data["mode"] >= 2:
            current_gnss_data = sl.GNSSData()
            current_gnss_data.set_coordinates(gpsd_data["lat"], gpsd_data["lon"], gpsd_data["altMSL"], False)
            current_gnss_data.longitude_std = 0.001
            current_gnss_data.latitude_std = 0.001
            current_gnss_data.altitude_std = 1.0

            position_covariance = [
                gpsd_data["eph"] * gpsd_data["eph"],
                0.0,
                0.0,
                0.0,
                gpsd_data["eph"] * gpsd_data["eph"],
                0.0,
                0.0,
                0.0,
                gpsd_data["epv"] * gpsd_data["epv"]
            ]
            current_gnss_data.position_covariances = position_covariance
            timestamp_microseconds = int(gpsd_data["time"].timestamp() * 1000000)
            ts = sl.Timestamp()
            ts.set_microseconds(timestamp_microseconds)
            current_gnss_data.ts = ts
            return current_gnss_data
        else:
            print("Fix lost : GNSS reinitialization")
            self.initialize()
            return None

    def grab(self):
        if self.new_data:
            self.new_data = False
            return sl.ERROR_CODE.SUCCESS, self.current_gnss_data
        return sl.ERROR_CODE.FAILURE, None

    def grabGNSSData(self):
        while self.continue_to_grab:
            with self.is_initialized_mtx:
                if self.is_initialized:
                    break
            time.sleep(0.001)

        while self.continue_to_grab:
            self.current_gnss_data = self.getNextGNSSValue()
            if self.current_gnss_data is not None:
                self.new_data = True

    def stop_thread(self):
        self.continue_to_grab = False
